### DISCUSSION TITLE
Gathering state callback invocation after creating data channel, communication channels opening rules.

### DISCUSSION THREAD

**Discussion by OlegDanko** [paullouisageneau/libdatachannel#857] [Q&A]
Hello
I am new to WebRTC, so I'm not confident about how it's supposed to work. I'd appreciate any help with understanding the following:

1. Consider the following code:
```
auto rtc_peer = std::make_shared<rtc::PeerConnection>();
rtc_peer->onGatheringStateChange([](auto state) {
    std::cout << state << std::endl;
});

std::this_thread::sleep_for(std::chrono::seconds(2));
std::cout << "Adding video track" << std::endl;
auto track = rtc_peer->addTrack(
    rtc::Description::Video("video", rtc::Description::Direction::SendOnly)
);

std::this_thread::sleep_for(std::chrono::seconds(2));
std::cout << "Opening data channel" << std::endl;
auto dc = rtc_peer->createDataChannel("ping-pong");

std::this_thread::sleep_for(std::chrono::milliseconds(200));
```
After running this code, I get the following output:
```
Adding video track
Opening data channel
in-progress
complete
```
It appears that creating a data channel causes the onGatheringStateChange callback to be invoked (whereas adding a track doesn't). Is this the way it is intended to work?

2. Consider the following code:
```
auto rtc_peer = std::make_shared<rtc::PeerConnection>();
rtc_peer->setLocalDescription();
auto description_before = rtc_peer->localDescription();

auto dc = rtc_peer->createDataChannel("ping-pong");

rtc_peer->setLocalDescription();
auto description_after = rtc_peer->localDescription();
```
After running this piece of code, description_before is set to an uninitialized optional, and description_after is initialized.
With this in mind, I have a few questions:
Is it possible/proper to formulate an RTC offer without any communication channels added? (If so, how can a local description of a peer without communication channels be accessed?)
Is it possible to open a communication channel after communication has been established?

**Comment by paullouisageneau** [comment]
> It appears that creating a data channel causes the onGatheringStateChange callback to be invoked (whereas adding a track doesn't). Is this the way it is intended to work?

Yes it is, by default `createDataChannel()` triggers the negotiation automatically by generating a local offer. The behavior can be changed be setting `rtc::Configuration::disableAutoNegotiation = true`, in this case nothing will happen until you call `setLocalDescription()`.

> After running this piece of code, description_before is set to an uninitialized optional, and description_after is initialized. 

You should not read the description at an arbitrary time as it is populated asynchronously. You should either use the `onLocalDescription` callback to get the description without ICE candidates (and get candidates when they are available), or the `onGatheringDone` callback to get the description with all the ICE candidates present.

> With this in mind, I have a few questions: Is it possible/proper to formulate an RTC offer without any communication channels added? (If so, how can a local description of a peer without communication channels be accessed?)

It is not possible with WebRTC, you have to negotiate something, either data channels or tracks.

> Is it possible to open a communication channel after communication has been established?

Yes you can open new data channels at any time, even after connection is completed. Of course data channels will only open when connected. Creating new data channels after the first one does not require new negotiation as they are negotiated in-band (creating new tracks requires it).
